### PR TITLE
Add release workflow to github, pypi, and docker hub

### LINKED ISSUE #2: Add release workflow

### PR DESCRIPTION
**Pull request by grqz**
Closes #2 

Examples (released by 81becb3):  
* Github release (plugin): [definitely-not-bgutil-ytdlp-pot-provider-to-distinguish-from-it 0.2.29](https://github.com/grqz/bgutil-ytdlp-pot-provider/releases/tag/0.2.29)
* PyPI release (plugin): [definitely-not-bgutil-ytdlp-pot-provider-to-distinguish-from-it 0.2.29](https://pypi.org/project/definitely-not-bgutil-ytdlp-pot-provider-to-distinguish-from-it/0.2.29/)
* Docker release (server): [grqz/bgutil-ytdlp-pot-provider](https://hub.docker.com/r/grqz/bgutil-ytdlp-pot-provider)

Known limitations:
- Can't automatically modify the PyPI project description
- Doesn't delete the tag before overwriting releases(apply https://stackoverflow.com/questions/5480258/how-can-i-delete-a-remote-tag to fix it if necessary)

Features:
* Creates Releases
* Also aims to replace build.yml
* Don't push to Docker Hub when run from `pull_request` trigger. To push the image, trigger the workflow manually with `workflow_dispatch`
* Can overwrite PyPI, Docker Hub, Github releases

PR todos:
- [x] Fix Docker tags/labels, determine whether the `Docker meta` step is necessary. Probably generate with shell instead.
- [x] Add changelog for Github Release
- [x] Add "Overview" for Docker Hub Release. See if possible to get it automated
- [x] Fix Docker tags for prs
- [x] Allow overwriting binaries for github releases
- [x] Allow overwriting binaries for PyPI releases

to deploy this workflow file to this repo:
- [x] Change `project.name` back to normal in `plugin/pyproject.toml` (needs changes in the repo)
- [x] Change `project.version` to the normal one in `plugin/pyproject.toml` (needs changes in the repo)
- [x] Follow [Publishing package distribution releases using GitHub Actions CI/CD workflows](https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/) to create a pending publisher under a PyPI account
- [x] Set repo secret (follow [Creating secrets for a repository](https://docs.github.com/en/actions/security-for-github-actions/security-guides/using-secrets-in-github-actions#creating-secrets-for-a-repository)) `DOCKERHUB_USERNAME`(docker hub username) and `DOCKERHUB_TOKEN`(docker hub password)

to make a new release:
1. Increment `project.version` in `plugin/pyproject.toml`
2. Trigger the workflow in [the actions page](https://github.com/Brainicism/bgutil-ytdlp-pot-provider/actions/workflows/release.yml)

Code References:
* https://github.com/coletdjnz/yt-dlp-get-pot/blob/master/.github/workflows/release.yml
* https://github.com/docker/build-push-action/blob/v5/README.md
* https://docs.github.com/en/actions/use-cases-and-examples/publishing-packages/publishing-docker-images
* https://github.com/docker/metadata-action/blob/master/README.md
* https://stackoverflow.com/a/5261470
* https://github.com/coletdjnz/yt-dlp-get-pot/blob/master/pyproject.toml
* https://stackoverflow.com/questions/64781462/github-actions-default-branch-variable

### CODE REVIEW

File: plugin/pyproject.toml
Comment (Brainicism) [inline plugin/pyproject.toml:7]: Just remember to fix this before we merge

File: .github/workflows/release.yml
Comment (Brainicism) [inline .github/workflows/release.yml:24]: The tags I've created so far are lightweight tags, which don't have a `taggerdate` associated with them. We can change this going forward with however we want to prepare releases, but sorting by `creatordate` instead will work for both lightweight and annotated tags. 

https://git-scm.com/docs/git-for-each-ref

File: .github/workflows/release.yml
Comment (Brainicism) [inline .github/workflows/release.yml:136]: There's a couple of this `event` checks, but this workflow is currently allowed for `workflow_dispatch`. Are you anticipating integrating this into the PR workflow in the future?

**Review by Brainicism — CHANGES_REQUESTED** [review]
Thanks for the awesome work! Just a few comments

File: .github/workflows/release.yml
Comment (grqz) [reply .github/workflows/release.yml:136]: No. Just copied from some document. This actually replaces that build.yml if we change the workflow `on`

**Comment by Brainicism** [comment]
Not sure if we want to do it in this PR or after the first pypi release, but we should include the pip install instructions on the README

File: plugin/pyproject.toml
Comment (grqz) [reply plugin/pyproject.toml:7]: yeah it's in the task list
> * [ ]  Change `project.name` back to normal in `plugin/pyproject.toml` (needs changes in the repo)
> * [ ]  Change `project.version` to the normal one in `plugin/pyproject.toml` (needs changes in the repo)

**Comment by grqz** [comment]
> Not sure if we want to do it in this PR or after the first pypi release, but we should include the pip install instructions on the README

probably do a follow-up pr after a pypi release is made

File: .github/workflows/release.yml
Comment (grqz) [reply .github/workflows/release.yml]: done in 33c17c1

**Comment by Brainicism** [comment]
I've set the required repo secrets for dockerhub personal access token

File: plugin/pyproject.toml
Comment (Brainicism) [inline plugin/pyproject.toml:15]: Can we add yt-dlp-get-pot as a dependency

**Comment by Brainicism** [comment]
Created trusted publisher on pypi

File: plugin/pyproject.toml
Comment (grqz) [reply plugin/pyproject.toml]: breaks github release workflow. trying to fix it

Comment (grqz) [reply plugin/pyproject.toml]: fixed

**Comment by grqz** [comment]
just built 0.2.16 and put links in the pr description. Any other issues?

if no then let's do this
> to deploy this workflow file to this repo:
> 
> * [ ]  Change `project.name` back to normal in `plugin/pyproject.toml` (needs changes in the repo)
> * [ ]  Change `project.version` to the normal one in `plugin/pyproject.toml` (needs changes in the repo)

**Comment by grqz** [comment]
are we going to use [attestations](https://github.com/actions/attest-build-provenance)?

File: plugin/pyproject.toml
Comment (grqz) [reply plugin/pyproject.toml]: Actually I'm not sure whether this(adding yt-dlp-get-pot to the dependencies) is expected. Because the upstream(yt-dlp/yt-dlp-sample-plugins#3) meant to pack the dependencies into the bundle, but imo we don't need to pack yt-dlp-get-pot into our zip. I simply removed the part of the script that pack dependencies. It's buggy.

Comment (Brainicism) [reply plugin/pyproject.toml]: My impression was just based off of: 
```
The provider plugin assumes this plugin is installed. You can define it as a Python dependency in your plugin package, or users can install it manually.
```

https://github.com/coletdjnz/yt-dlp-get-pot?tab=readme-ov-file#developing-a-provider-plugin 

Might be worth raising with cole but not a blocker for this PR

**Comment by Brainicism** [comment]
> just built 0.2.16 and updated links in the pr description. Any other issues?
> 
> if no then let's do this
> 
> > to deploy this workflow file to this repo:
> > 
> > * [ ]   Change `project.name` back to normal in `plugin/pyproject.toml` (needs changes in the repo)
> > * [ ]   Change `project.version` to the normal one in `plugin/pyproject.toml` (needs changes in the repo)

No other issues, good to approve after those changes

File: .github/workflows/release.yml
Comment (Brainicism) [inline .github/workflows/release.yml:50]: The syntax highlighting breaks here. It looks fine to me, but you've made sure that this part works right?

Comment (grqz) [reply .github/workflows/release.yml:50]: Use github.dev to see syntax highlighting. It is working, see
> Examples (released by [f19300f](https://github.com/Brainicism/bgutil-ytdlp-pot-provider/commit/f19300fe510a19156a7e270cfe46524d8d2baa83)):
> 
> * Github release (plugin): [definitely-not-bgutil-ytdlp-pot-provider-to-distinguish-from-it v0.2.16](https://github.com/grqz/bgutil-ytdlp-pot-provider/releases/tag/v0.2.16)

The changelog currently looks like this. Do we need to change it to the one like in your repo? It's cumbersome to get github user login name in this script

File: plugin/pyproject.toml
Comment (grqz) [reply plugin/pyproject.toml]: alright let's resolve this conversation

File: .github/workflows/release.yml
Comment (Brainicism) [reply .github/workflows/release.yml:50]: I think it's fine, not a huge deal with the usernames aren't there for now

Comment (grqz) [reply .github/workflows/release.yml:50]:
```sh
gh pr list --json author,number,title -q '.[]| "* \(.title) " + "by \(.author.login)" + " (#\(.number|tostring))"' -R Brainicism/bgutil-ytdlp-pot-provider --state merged
```
this should work with `--since` and the last tag's creatordate

Comment (grqz) [reply .github/workflows/release.yml]: Now it looks like the one in your repo
v0.2.24: https://github.com/grqz/bgutil-ytdlp-pot-provider/releases/tag/v0.2.24

File: .github/workflows/release.yml
Comment (grqz) [inline .github/workflows/release.yml:32]: Do you want to keep the git tag naming style in your repo(like `0.2.1`) or do you want the one with prefix v(like `v0.2.1`)?

Comment (Brainicism) [reply .github/workflows/release.yml:32]: Let's keep it without the prefix

Comment (grqz) [reply .github/workflows/release.yml]: done in 0696592 (not yet tested)

**Comment by grqz** [comment]
Included pr workflow
see https://github.com/grqz/bgutil-ytdlp-pot-provider/actions/runs/10661455258/job/29546966925?pr=9

**Comment by grqz** [comment]
> * [x]  Change `project.name` back to normal in `plugin/pyproject.toml` (needs changes in the repo)
> * [x]  Change `project.version` to the normal one in `plugin/pyproject.toml` (needs changes in the repo)

done in 62aa290

File: .github/workflows/build.yml
Comment (Brainicism) [inline .github/workflows/build.yml:1]: Is this intent to run `release.yml` on every PR as well?

Comment (grqz) [reply .github/workflows/build.yml:1]: yes, it works in the same way as the original build.yml which has been deleted
example: https://github.com/grqz/bgutil-ytdlp-pot-provider/actions/runs/10661782621/job/29547857638?pr=10

File: .github/workflows/release.yml
Comment (Brainicism) [inline .github/workflows/release.yml:44]: Can we move this conditional to the job level so we can skip the entire job instead?

Comment (grqz) [reply .github/workflows/release.yml:44]: No. The docker job currently `needs` the job.

**Review by Brainicism — APPROVED** [review]

File: plugin/pyproject.toml
Comment (grqz) [inline plugin/pyproject.toml:10]:
```suggestion
keywords = ["yt-dlp", "yt-dlp-plugin", "yt-dlp-plugins-get-pot"]
```
maybe add a keyword?

**Comment by Brainicism** [comment]
Good to merge at your discretion

**Comment by grqz** [comment]
> Good to merge at your discretion

all things on the TODO list have been done, you can merge it now

**Pull request merged**
